File: easy/101-200/116/rust/src/main.rs
// Copyright 2022 CJ Harries
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use itertools::Itertools;

fn main() {
    println!("rad");
}

fn permute_string(input: &str) -> Vec<String> {
    let mut result = Vec::new();
    let chars = input.chars().collect::<Vec<char>>();
    for permutation in (0..chars.len()).permutations(chars.len()) {
        let mut permutation_string = String::new();
        for i in permutation {
            permutation_string.push(chars[i]);
        }
        result.push(permutation_string);
    }
    result
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_permute_string() {
        assert_eq!(permute_string("ab"), vec!["ab", "ba"]);
        assert_eq!(
            permute_string("abc"),
            vec!["abc", "acb", "bac", "bca", "cab", "cba"]
        );
    }
}

File: easy/101-200/116/go/main.go
// Copyright 2021 CJ Harries
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"fmt"
	"sort"
)

var zPrint = fmt.Println

func main() {
	_, _ = zPrint(getAllUniqueStringPermutations("abbccc"))
}

func getAllStringPermutations(stringToPermute string, startingIndex int) (permutations []string) {
	if len(stringToPermute) <= startingIndex {
		return []string{stringToPermute}
	}
	stringAsRunes := []rune(stringToPermute)
	for movingIndex := startingIndex; movingIndex < len(stringAsRunes); movingIndex++ {
		stringAsRunes[startingIndex], stringAsRunes[movingIndex] = stringAsRunes[movingIndex], stringAsRunes[startingIndex]
		permutations = append(permutations, getAllStringPermutations(string(stringAsRunes), startingIndex+1)...)
		stringAsRunes[startingIndex], stringAsRunes[movingIndex] = stringAsRunes[movingIndex], stringAsRunes[startingIndex]
	}
	return
}

func getAllUniqueStringPermutations(stringToPermute string) (permutations []string) {
	nonuniquePermutations := getAllStringPermutations(stringToPermute, 0)
	permutationMap := make(map[string]bool, len(nonuniquePermutations))
	for _, permutation := range nonuniquePermutations {
		permutationMap[permutation] = true
	}
	for key := range permutationMap {
		permutations = append(permutations, key)
	}
	sort.Strings(permutations)
	return
}
